File: core/lib/glob.ts
/**
 * Tiny glob → RegExp converter. Supports the subset torba uses across
 * `restrict` sweeping and `pipe` artifact selectors:
 *
 *   `*`        — any sequence of non-separator chars
 *   `**`       — any sequence of chars including separators (zero or more
 *                path segments when surrounded by `/`)
 *   `?`        — exactly one non-separator char
 *   `{a,b,c}`  — flat alternation (no nesting)
 *
 * Separator is `/` regardless of platform; callers normalize Windows
 * backslashes before matching.
 */

const STAR = '__TORBA_GLOBSTAR_';
const RX_META = /[.+^$()|[\]\\]/;
const ALT_META = /[.+^$()|[\]\\*?{}]/g;

export function globToRegex(glob: string): RegExp {
  let s = glob;
  // Distinct sentinels for the four globstar positions, processed in
  // most-specific-first order so they don't overlap.
  s = s.replace(/\/\*\*$/, `\0${STAR}END\0`); //   …/**       (suffix)
  s = s.replace(/\/\*\*\//g, `\0${STAR}MID\0`); // …/**/…     (middle)
  s = s.replace(/^\*\*\//, `\0${STAR}START\0`); //   **/…     (prefix)
  s = s.replace(/\*\*/g, `\0${STAR}BARE\0`); //     **        (bare)

  let out = '';
  let i = 0;
  while (i < s.length) {
    const ch = s[i]!;
    if (ch === '\0') {
      const close = s.indexOf('\0', i + 1);
      const tag = s.slice(i + 1 + STAR.length, close);
      if (tag === 'END') out += '(?:/.*)?';
      else if (tag === 'MID') out += '(?:/.*)?/';
      else if (tag === 'START') out += '(?:.*/)?';
      else out += '.*'; // BARE
      i = close + 1;
    } else if (ch === '*') {
      out += '[^/]*';
      i += 1;
    } else if (ch === '?') {
      out += '[^/]';
      i += 1;
    } else if (ch === '{') {
      const end = s.indexOf('}', i);
      if (end === -1) {
        out += '\\{';
        i += 1;
      } else {
        const parts = s.slice(i + 1, end).split(',');
        out +=
          '(?:' + parts.map((p) => p.replace(ALT_META, '\\$&')).join('|') + ')';
        i = end + 1;
      }
    } else if (RX_META.test(ch)) {
      out += '\\' + ch;
      i += 1;
    } else {
      out += ch;
      i += 1;
    }
  }
  return new RegExp('^' + out + '$');
}

/**
 * The longest prefix of `glob` that contains no glob metacharacters,
 * truncated to the last path separator. Used as the filesystem starting
 * point for sweep-walking.
 *
 * `/home/x/mods/**\/*.jar` → `/home/x/mods`
 * `/home/x/mods/foo.jar`   → `/home/x/mods`
 * `*.jar`                   → `''` (caller substitutes cwd)
 */
export function globBase(glob: string): string {
  let lastSlash = -1;
  for (let i = 0; i < glob.length; i++) {
    const ch = glob[i]!;
    if (ch === '*' || ch === '?' || ch === '{' || ch === '[') break;
    if (ch === '/') lastSlash = i;
  }
  return lastSlash >= 0 ? glob.slice(0, lastSlash) : '';
}

File: crates/opys-core/src/glob.rs
//! Tiny glob → Regex converter. Frozen semantics — see TS `core/lib/glob.ts`.
//!
//! `*` non-separator chars, `**` any chars, `?` one non-sep, `{a,b}` flat alt.
//! Separator is `/`; callers normalize Windows backslashes.

use regex::Regex;

const STAR: &str = "__OPYS_GLOBSTAR_";

fn is_rx_meta(c: char) -> bool {
    matches!(c, '.' | '+' | '^' | '$' | '(' | ')' | '|' | '[' | ']' | '\\')
}

fn is_alt_meta(c: char) -> bool {
    matches!(
        c,
        '.' | '+' | '^' | '$' | '(' | ')' | '|' | '[' | ']' | '\\' | '*' | '?' | '{' | '}'
    )
}

/// Compile a glob to an anchored `Regex`. Pattern source is `^…$`.
pub fn glob_to_regex(glob: &str) -> Regex {
    let mut s = glob.to_owned();
    let null = '\0';
    // Most-specific-first so the sentinels don't overlap.
    s = replace_pattern(&s, r"/\*\*$", &format!("{null}{STAR}END{null}"));
    s = replace_pattern(&s, r"/\*\*/", &format!("{null}{STAR}MID{null}"));
    s = replace_pattern(&s, r"^\*\*/", &format!("{null}{STAR}START{null}"));
    s = replace_pattern(&s, r"\*\*", &format!("{null}{STAR}BARE{null}"));

    let mut out = String::with_capacity(s.len() * 2);
    let chars: Vec<char> = s.chars().collect();
    let mut i = 0;
    while i < chars.len() {
        let ch = chars[i];
        if ch == '\0' {
            let close = chars[i + 1..]
                .iter()
                .position(|&c| c == '\0')
                .map(|p| p + i + 1)
                .expect("matching sentinel");
            let tag: String = chars[i + 1 + STAR.len()..close].iter().collect();
            out.push_str(match tag.as_str() {
                "END" => "(?:/.*)?",
                "MID" => "(?:/.*)?/",
                "START" => "(?:.*/)?",
                _ => ".*", // BARE
            });
            i = close + 1;
        } else if ch == '*' {
            out.push_str("[^/]*");
            i += 1;
        } else if ch == '?' {
            out.push_str("[^/]");
            i += 1;
        } else if ch == '{' {
            if let Some(end) = chars[i..].iter().position(|&c| c == '}') {
                let end = end + i;
                let parts: String = chars[i + 1..end].iter().collect();
                let alts: Vec<String> = parts
                    .split(',')
                    .map(|p| {
                        let mut esc = String::with_capacity(p.len() * 2);
                        for c in p.chars() {
                            if is_alt_meta(c) {
                                esc.push('\\');
                            }
                            esc.push(c);
                        }
                        esc
                    })
                    .collect();
                out.push_str("(?:");
                out.push_str(&alts.join("|"));
                out.push(')');
                i = end + 1;
            } else {
                out.push_str("\\{");
                i += 1;
            }
        } else if is_rx_meta(ch) {
            out.push('\\');
            out.push(ch);
            i += 1;
        } else {
            out.push(ch);
            i += 1;
        }
    }
    Regex::new(&format!("^{out}$")).expect("valid glob → regex")
}

fn replace_pattern(haystack: &str, pattern: &str, replacement: &str) -> String {
    Regex::new(pattern).unwrap().replace_all(haystack, replacement).into_owned()
}

/// Longest non-glob prefix, truncated to the last `/`.
pub fn glob_base(glob: &str) -> String {
    let mut last_slash: Option<usize> = None;
    for (i, ch) in glob.char_indices() {
        if matches!(ch, '*' | '?' | '{' | '[') {
            break;
        }
        if ch == '/' {
            last_slash = Some(i);
        }
    }
    match last_slash {
        Some(i) => glob[..i].to_owned(),
        None => String::new(),
    }
}
